django-rgd/rgd/models/common.py

```diff
@@ -226,13 +228,39 @@ def download_to_local_path(self, directory: str):
         """Forcibly download this file to a directory on disk.
 
         Cleanup must be handled by caller.
+
+        This will handle thread locking to prevent multiple processes/threads
+        from trying to download the file at the same time -- only one thread
+        will perfrom the download and the rest will yield the its result.
+
         """
-        if self.type == FileSourceType.FILE_FIELD:
-            return download_field_file_to_local_path(self.file, directory, self.name)
-        elif self.type == FileSourceType.URL:
-            return download_url_file_to_local_path(self.url, directory, self.name)
+        # Thread/process safe locking for file access
+        dest_path = Path(directory, self.name)
+        dest_path.parent.mkdir(parents=True, exist_ok=True)
+        lock_file_path = f'{dest_path}.lock'
+        lock = FileLock(
+            lock_file_path
+        )  # timeout=getattr(settings, 'RGD_FILE_LOCK_TIMEOUT', 10000))
+
+        with lock:  # TODO: handle timeouts in condition
+            # Files older than 7 days will be re-downloaded
+            if (
+                dest_path.exists()
+                and dest_path.stat().st_size > 0
+                and (datetime.now() - datetime.fromtimestamp(os.path.getmtime(dest_path))).days
+                <= getattr(settings, 'RGD_FILE_CACHE_MAX_DAYS', 7)
+            ):
+                # File already exists (is cached)
+                logger.info(f'Found cached file ({self.pk}) at: {dest_path}')
```

Comment:
I'm actually attempting to use `getattr(settings, 'RGD_FILE_CACHE_MAX_DAYS', 7)` here to re-download the file if has been sitting around any longer than `RGD_FILE_CACHE_MAX_DAYS`. This is for the rare case in which a file may have changed, so that we do not continue to use an outdated resource.

Considering that we may care more about memory management and need to clean up files regularly, this scenario may never happen. In which case, `touch`ing the files to know how recently they were accessed and cleaning up ould files to make room for new files would be best.

I will try to implement this.